### PR TITLE
[lld][COFF][NFC] Factor out exception table sorting.

### PR DESCRIPTION
**Pull request by cjacek**
This is a preparation for ARM64EC support, which needs to sort both ARM and x86_64 tables separately.

### CODE REVIEW

File: lld/COFF/Writer.cpp [hunk 1]
```diff
@@ -2164,41 +2165,49 @@ void Writer::writeBuildId() {
...
+  llvm::TimeTraceScope timeScope("Sort exception table");
+
+  struct EntryX64 {
+    ulittle32_t begin, end, unwind;
+  };
+  struct EntryArm {
+    ulittle32_t begin, unwind;
+  };
+
   if (ctx.config.machine == AMD64) {
```
Comment (compnerd) [inline lld/COFF/Writer.cpp:2201]: I wonder if this is better suited to be a switch statement.

Comment (cjacek) [reply lld/COFF/Writer.cpp]: Good idea, it makes things cleaner, thanks!

File: lld/COFF/Writer.cpp [hunk 2]
```diff
@@ -2164,41 +2165,49 @@ void Writer::writeBuildId() {
...
-  uint8_t *begin = bufAddr(firstPdata);
-  uint8_t *end = bufAddr(lastPdata) + lastPdata->getSize();
+  uint8_t *begin = bufAddr(first);
+  uint8_t *end = bufAddr(last) + last->getSize();
+  if ((end - begin) % sizeof(T) != 0) {
+    fatal("unexpected .pdata size: " + Twine(end - begin) +
+          " is not a multiple of " + Twine(sizeof(T)));
+  }
+
+  parallelSort(MutableArrayRef<T>((T *)begin, (T *)end),
```
[the remark this answers is not in the document]
Comment (cjacek) [reply lld/COFF/Writer.cpp]: Done, thanks.

**Review by compnerd — APPROVED** [review]

**Review by rnk — APPROVED** [review]

**Comment by cjacek** [comment]
Thanks for reviews. I merged with a small fixup to check firstPdata earlier. Previous version was fine in practice, but I think it could upset sanitizers because of uninitialized lastPdata being passed.

**Pull request merged**